tests/unit/test_geometry/test_battery_geometry.py

```diff
@@ -1,28 +1,30 @@
 #
 # Tests for the base model class
 #
+import pytest
 
 import pybamm
-import unittest
 
 
-class TestBatteryGeometry(unittest.TestCase):
-    def test_geometry_keys(self):
-        for cc_dimension in [0, 1, 2]:
-            geometry = pybamm.battery_geometry(
-                options={
-                    "particle size": "distribution",
-                    "dimensionality": cc_dimension,
-                },
+class TestBatteryGeometry:
+    @pytest.fixture(params=[0, 1, 2])
+    def geometry(self, request):
+        geometry = pybamm.battery_geometry(
+            options={
+                "particle size": "distribution",
+                "dimensionality": request.param,
+            },
+        )
+        return geometry
+
+    def test_geometry_keys(self, geometry):
+        for domain_geoms in geometry.values():
+            assert all(
+                isinstance(spatial_var, str) for spatial_var in domain_geoms.keys()
             )
-            for domain_geoms in geometry.values():
-                all(
-                    self.assertIsInstance(spatial_var, str)
-                    for spatial_var in domain_geoms.keys()
-                )
         geometry.print_parameter_info()
 
-    def test_geometry(self):
+    def test_geometry(self, geometry):
```

Comment:
We initialise the `geometry` variable below, do we need it in the arguments of the function?

Reply:
Also, this test can be refactored since we are testing the same thing repeatedly. Please add this to the tracking issue.

Reply:
Now sure how to refactor. Could you help me more?

Reply:
Usually, in such situations, it's better to make the test shorter and distribute it into multiple tests. For example, here, you could rewrite `test_geometry` into an array of smaller tests (haven't added the `self` keyword here for brevity):

```python
def test_battery_geometry_current_collector_tabs():
    geometry = pybamm.battery_geometry(
        options={
            "particle size": "distribution",
            "dimensionality": 1,
        },
    )
    assert "tabs" in geometry["current collector"]
```

and

```python
def test_battery_geometry_cylindrical_dimensionality():
    geo = pybamm.geometric_parameters
    geometry = pybamm.battery_geometry(
        form_factor="cylindrical", options={"dimensionality": 1}
    )
    assert geometry["current collector"]["r_macro"]["min"] == geo.r_inner
    assert geometry["current collector"]["r_macro"]["max"] == 1
```

However, readability is important. So, if refactoring introduces too many conditions or makes the code hard to read, it is best to not bother and keep the test as is.

Another aspect is to remove the loops by parametrizing. For example, the loop `for cc_dimension in [0, 1, 2]:` can be refactored into

```python
@pytest.mark.parametrize("cc_dimension", [0, 1, 2])
def test_battery_geometry_basics(cc_dimension):
```

and similarly

```python
@pytest.mark.parametrize("electrode, particle_type", [
    ("negative", "primary"),
    ("negative", "secondary"),
    ("positive", "primary"),
    ("positive", "secondary"),
])
def test_battery_geometry_particle_phases(electrode, particle_type):
    options = {"particle phases": "2"}
    geometry = pybamm.battery_geometry(options=options)
    geo = pybamm.GeometricParameters(options=options)
# followed by the rest of the test
```

and so on. I hope this can get you going!

Reply:
I'm happy to write more examples if needed, it's easier than you would think :)

Reply:
These examples looks enough to me. Thanks :)
I guess we don't need case or if/else then as we using `parametrize` marker?

Reply:
In some cases, we might still need them.

```python

@pytest.mark.parametrize("foo, bar", [0, 1, 2, 3, 4])
def test_something(foo, bar):
    # Some logic for initialising "foo" and "bar" here
    # and setting attributes, etc.
    # [...]

    # Conditional logic
    match foo.value:
        case 1:
            # do something
            assert foo in bar
        case 3.value:
            # do something else
            assert bar in foo

    # not (1) and not (3) – i.e., for (0) or (2) or (4)
    # Some other logic that applies to the rest of the cases
    assert foo.value + bar.value > 30
```

but it depends on the test. It _might_ be better to handle such cases without parameterizing them